Feature 1 and routing

AndreeaGui opened the pull request:
I implemented feature 1 and react routing

Commit:
Feature 1 and routing

Commit:
Second commit

contains feature 1 and react routing - the same as the previous commit but differently put on Git

serban-petrescu commented on src/App.js:
```diff
@@ -0,0 +1,28 @@
+import SmartQuestionsList from './view/SmartQuestionsList';
+import SmartCreateQuestion from './view/SmartCreateQuestion';
+import SmartFilterQuestionsByTag from './view/SmartFilterQuestionsByTag';
+import SmartFilterQuestionsByTtitle from './view/SmartFilterQuestionsByTitle';
```
Typo (Ttitle).

AndreeaGui replied:
thank you, this saves some of my future time

serban-petrescu commented on src/view/CreateQuestion.js:
```diff
@@ -0,0 +1,27 @@
+        <h2>Add question</h2>
+        <div>
+            <label>User ID: </label>
+            <input value={newUserId} onChange={ e =>onChange("userId",e.target.value)}/>
```
Alt + Shift + F is your friend. Get used to use it to format your code. Trust me, it will really help you later, especially at the beginning of your career...

AndreeaGui replied:
You told me the advice for formatting at the previous assignment too and I did it at the beginning of this one, but I forgot meanwhile. Thank you for the reminder!

serban-petrescu commented on src/view/Router.js:
```diff
@@ -0,0 +1,15 @@
+import SmartCreateQuestion from "./SmartCreateQuestion";
+import SmartQuestionDetails from "./SmartQuestionDetails";
+
+const Router = ({route}) => (
```
As you are using the react router now, you can safely remove this and the smart router.

serban-petrescu commented on src/view/QuestionDetails.js:
```diff
@@ -0,0 +1,25 @@
+import React from "react"
+
+const SudentDetails = ({ userId, title, text, tags, index }) => (
```
You forgot to adjust the name of this class :)

serban-petrescu commented on src/view/QuestionsList.js:
```diff
@@ -0,0 +1,49 @@
+            <tbody>
+            {
+                questions.map((question, index) => (
+                    <tr key = {index}>
```
As you have this table duplicated between the QuestionList and the FilterBy... components, I would suggest that you extract a dumb component out of the table.

AndreeaGui replied:
Yes, I will try to do this and use that dumb component every time when I have to display a list of questions.

serban-petrescu commented on src/view/QuestionsList.js:
```diff
@@ -0,0 +1,49 @@
+        <h2>Questions</h2>
+        <br/> <br/> 
+        <label>Search: </label>
+            <input //value={search} onSearch={ e =>onSearch("userId",e.target.value)}
```
I guess that the search does not work yet (also judging by the content of the FilterBy... components).

serban-petrescu reviewed (commented):
## Grading

Note that you must present at the laboratory one operation out of each feature (to prove that it works) and answer any questions from me related to your code (during the lab):
 1. Create a question and then list all of them.
 
If you are unable to show that the assignment works or answer my questions, this assignment will not be considered as being handed-in.

### Minimum requirements for a passing grade:
 * [X] Browser-based U.I.,
 * [X] Feature 1,
 * [X] Clear MV* separation,
 * [X] Observable model.

### Additional requirements
 * [X] Primitive model-based routing (6)
 * [ ] Feature 2 (8)
 * [ ] Feature 3 (10)

### Bonus features
 * [ ] Bonus feature 1
 * [ ] Bonus feature 2
  
### Bonus requirements
 * [ ] Prepare and present mockups (+1.0)
 * [ ] Use Redux for holding the state (+2.0)
 * [X] Use React Router for handling the routing (+1.0)
 * [ ] Use a CSS library for styling (+1.0)
 * [ ] Write component tests for your views (+1.0)

**Final grade: 5.5.** 

### Notes
 - **minus 1.5 points: Feature 1 is partial**: Your filtering components + input field do not exist / work yet (you are not handling the input field value, nor are you doing a round-trip to the model). I guess this might be a sligh miss-comunication between the two of us. To clarify, I expect to see that your input fields / buttons / whatever components for filtering exist, the values and states of these components are stored in the model (e.g. for filter by title, you will have an input field, whose value should be stored in the model) and that when pressing a "search" button the displayed questions are **updated** somehow (it can be hard-coded or you can do a simple filtering using the `filter` function, however you want). Most students actually implemented the filtering also on the UI just to see that it works.

Commit:
Features + CSS

Features 1, 2 are working. Feature 3 has the skeleton and will be completely implemented in assignment 3.

serban-petrescu commented on src/App.js:
```diff
@@ -2,23 +2,25 @@ import React from 'react';
 
 
-import {HashRouter, Switch, Route} from "react-router-dom"
+import { HashRouter, Switch, Route } from "react-router-dom"
```
Did you use the format button on the IDE? :O Nice!

serban-petrescu commented on src/model/answer.js:
```diff
@@ -0,0 +1,109 @@
+    addAnswer(userId, questionId, text) {
+        this.state = {
+            ...this.state,
+            currentId: ++this.state.currentId,
```
++ modifies the value in place btw :) The whole purpose of creating a new object was to not change the old once (so you should simply do +1 and not ++).

serban-petrescu commented on src/view/SmartCreateQuestion.js:
```diff
@@ -11,29 +11,29 @@ const mapModelStateToComponentState = modelState => ({
-    componentWillUnmount(){
+    componentWillUnmount() {
         question.removeListener("change", this.listener);
         //prevent memory leacks
```
Almost correct spelling.

serban-petrescu reviewed (commented):
## Grading

### Minimum requirements for a passing grade:
 * [X] Browser-based U.I.,
 * [X] Feature 1,
 * [X] Clear MV* separation,
 * [X] Observable model.

### Additional requirements
 * [X] Primitive model-based routing (6)
 * [X] Feature 2 (8)
 * [X] Feature 3 (10)


### Bonus requirements
 * [ ] Prepare and present mockups (+1.0)
 * [ ] Use Redux for holding the state (+2.0)
 * [X] Use React Router for handling the routing (+1.0)
 * [X] Use a CSS library for styling (+1.0)
 * [ ] Write component tests for your views (+1.0)

**Final grade: 10.** 

### Notes
 - Very good job! I am really happy about your progress.
